mmcv/ops/csrc/pytorch/cuda/filtered_lrelu.cu

```diff
@@ -0,0 +1,1879 @@
+#define MAX(A, B) ((A) > (B) ? (A) : (B))
+#define CEIL_DIV(A, B)                                   \
+  (((B) == 1)                                            \
+       ? (A)                                             \
+       : ((B) == 2) ? ((int)((A) + 1) >> 1)              \
+                    : ((B) == 4) ? ((int)((A) + 3) >> 2) \
+                                 : (((A) + ((A) > 0 ? (B)-1 : 0)) / (B)))
+
+// This works only up to blocks of size 256 x 256 and for all N that are powers
+// of two.
+template <int N>
+__device__ __forceinline__ void fast_div_mod(int& x, int& y, unsigned int i) {
+  if ((N & (N - 1)) && N <= 256)
+    y = (i * ((1 << 24) / N + 1)) >> 24;  // Assumes N <= 256, i < N*256.
+  else
+    y = i / N;
+
+  x = i - y * N;
+}
+
+// Type cast stride before reading it.
+template <class T>
+__device__ __forceinline__ T get_stride(const int64_t& x) {
+  return *reinterpret_cast<const T*>(&x);
+}
+
+//------------------------------------------------------------------------
+// Filters, setup kernel, copying function.
+
+#define MAX_FILTER_SIZE 32
+
+// Combined up/down filter buffers so that transfer can be done with one copy.
+__device__ float
+    g_fbuf[2 * MAX_FILTER_SIZE * MAX_FILTER_SIZE];  // Filters in global memory,
+                                                    // written by setup kernel.
+__device__ __constant__ float
+    c_fbuf[2 * MAX_FILTER_SIZE *
+           MAX_FILTER_SIZE];  // Filters in constant memory, read by main
+                              // kernel.
+
+// Accessors to combined buffers to index up/down filters individually.
+#define c_fu (c_fbuf)
+#define c_fd (c_fbuf + MAX_FILTER_SIZE * MAX_FILTER_SIZE)
+#define g_fu (g_fbuf)
+#define g_fd (g_fbuf + MAX_FILTER_SIZE * MAX_FILTER_SIZE)
+
+// Set up filters into global memory buffer.
+static __global__ void setup_filters_kernel(filtered_lrelu_kernel_params p) {
+  for (int idx = threadIdx.x; idx < MAX_FILTER_SIZE * MAX_FILTER_SIZE;
+       idx += blockDim.x) {
+    int x, y;
+    fast_div_mod<MAX_FILTER_SIZE>(x, y, idx);
+
+    int fu_x = p.flip ? x : (p.fuShape.x - 1 - x);
+    int fu_y = p.flip ? y : (p.fuShape.y - 1 - y);
+    if (p.fuShape.y > 0)
+      g_fu[idx] = (x >= p.fuShape.x || y >= p.fuShape.y)
+                      ? 0.0f
+                      : p.fu[fu_x * p.fuStride.x + fu_y * p.fuStride.y];
+    else
+      g_fu[idx] =
+          (x >= p.fuShape.x || y > 0) ? 0.0f : p.fu[fu_x * p.fuStride.x];
+
+    int fd_x = p.flip ? x : (p.fdShape.x - 1 - x);
+    int fd_y = p.flip ? y : (p.fdShape.y - 1 - y);
+    if (p.fdShape.y > 0)
+      g_fd[idx] = (x >= p.fdShape.x || y >= p.fdShape.y)
+                      ? 0.0f
+                      : p.fd[fd_x * p.fdStride.x + fd_y * p.fdStride.y];
+    else
+      g_fd[idx] =
+          (x >= p.fdShape.x || y > 0) ? 0.0f : p.fd[fd_x * p.fdStride.x];
+  }
+}
+
+// Host function to copy filters written by setup kernel into constant buffer
+// for main kernel.
+template <bool, bool>
```

Comment:
It seems that the template arguments has never been used? Why do we need to define this function as a template?

Reply:
Seems used in https://github.com/open-mmlab/mmcv/blob/eff99f8a8841c59b47c263c97d6d64c35b2b9bb4/mmcv/ops/csrc/pytorch/cuda/filtered_lrelu.cu#L1752

Reply:
It will generate the same function instance whatever the argument is.